Updated migration

lucassaldanha opened the pull request.

jframe committed:
Privacy multitenancy validation (#296)

Signed-off-by: Jason Frame <jasonwframe@gmail.com>

usmansaleem committed:
[BESU-77] - Enable TLS for JSON-RPC HTTP Service (#271)

Expose following new command line parameters to enable TLS on Ethereum JSON-RPC HTTP interface to allow clients like Ethsigner to connect via TLS

--rpc-http-tls-enabled=true
(Optional - Only required if --rpc-http-enabled is set to true) Set to ‘true’ to enable TLS. false by default.

--rpc-http-tls-keystore-file="/path/to/cert.pfx"
(Must be specified if TLS is enabled) Path to PKCS12 format key store which contains server's certificate and it's private key

--rpc-http-tls-keystore-password-file="/path/to/cert.passwd"
(Must be specified if TLS is enabled) Path to the text file containing password for unlocking key store.

--rpc-http-tls-known-clients-file="/path/to/rpc_tls_clients.txt"
(Optional) Path to a plain text file containing space separated client’s certificate’s common name and its sha-256 fingerprints when they are not signed by a known CA. The presence of this file (even empty) will enable TLS client authentication i.e. the client will present its certificate to server on TLS handshake and server will establish that the client’s certificate is either signed by a proper/known CA otherwise server trusts client's certificate by reading it's sha-256 fingerprint from known clients file specified above. The format of the file is (as an example):

localhost DF:65:B8:02:08:5E:91:82:0F:91:F5:1C:96:56:92:C4:1A:F6:C6:27:FD:6C:FC:31:F2:BB:90:17:22:59:5B:50

Signed-off-by: Usman Saleem <usman@usmans.info>

shemnon committed:
Fix Gradle 7.0 compatibility issues and build warnings (#288)

Removes as many Gradle 7.0 compatibility issues as possible
* `baseName` -> `archiveBaseName`
* `extension` -> `archiveExtension`
* `destinationDir` -> `destinationDirectory`
* `runtime` -> `runtimeOnly`
* Change some log4j-api and log4j-core dependencies
* Remove an unneeded and outdated plugin (`net.ltgt.apt`)
* tweak the plugin-api change detector's property annotations.

Warnings still exist with one external plugin used for license file
checking that we do not control the source code for.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

CjHare committed:
Debug logging on Private Transaction failure (#302)

Signed-off-by: Christopher Hare <chris.hare@consensys.net>

AbdelStark committed:
Handling of various opcodes that generate message calls. (#300)

* Re-enable `TRACE` API.
* - Fixed broken tests. Tests were not triggering.
- Handle `0x` special case for `Store` class.
- Add test covering `CALLCODE` opcode

* Added `DELEGATECALL` test case.

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

RatanRSur committed:
eip 2124 line to changelog (#305)



Signed-off-by: Ratan Rai Sur <ratan.r.sur@gmail.com>

shemnon committed:
Changelog updates for Tuwneni (#306)

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

MadelineMurray committed:
Added priv_call to changelog (#307)

Signed-off-by: Madeline <madeline.murray@consensys.net>

EdJoJob committed:
1.4.0-beta1 release (#308)

Signed-off-by: {name} <{email}>

EdJoJob committed:
Prepare next release (#309)

Signed-off-by: Edward Evans <edward.joshua.evans@gmail.com>

shemnon committed:
TotalDifficulty is a BigInteger (#310)

Don't try and cast total difficulty down to a long because it will
overflow long in a reasonable timeframe (mainnet already has).

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

AbdelStark committed:
[Tracing API] Handling of multiple transactions in a block. (#304)

* Handle multiple transactions in the same block.
* Fix a bug in `BlockTracer`. The bug was causing multiple `TransactionTrace` to share the same list of `TraceFrame`.
A new method has been added in `DebugOperationTracer` to reset the list of frames for successive transactions in a block.

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

usmansaleem committed:
Add missing bouncy castle library required by JSON-RPC HTTP TLS (#313)

-- This jar is required by Tuweni API handling of known clients fingerprints at runtime.

Signed-off-by: Usman Saleem <usman@usmans.info>

MadelineMurray committed:
1.4.0-beta2 release (#312)

Signed-off-by: {name} <{email}>

EdJoJob committed:
Prepare for version 1.4.0-beta3-SNAPSHOT (#315)

Signed-off-by: {name} <{email}>

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/migration/PrivateStorageMetadataMigration.java:
```diff
@@ -0,0 +1,290 @@
+      final Hash blockHash = block.getHash();
+      final BlockHeader blockHeader = block.getHeader();
+      LOG.trace(
+          "Processing block {} ({}/{})", blockHash, blockNumber + 1, chainHeadBlockNumber + 1);
```
I think the blockNumber + 1 and chainHeadBlockNumber +1 could be confusing ...

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/migration/PrivateStorageMetadataMigration.java:
```diff
@@ -0,0 +1,290 @@
+        final Transaction pmt = pmtsInBlock.get(pmtIndex);
+        LOG.trace("Processing PMT {} ({}/{})", pmt.getHash(), pmtIndex + 1, pmtsInBlock.size());
+
+        final ReceiveResponse receiveResponse =
```
In real life you might not get a PTx for every PMT ... If that is the case you just start processing the next PMT :-)

iikirilov replied:
Yes, if it is not a privacy group you are in you will not have PTx for that PMT

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/migration/PrivateStorageMetadataMigration.java:
```diff
@@ -0,0 +1,290 @@
+                .orElse(PrivacyGroupHeadBlockMap.EMPTY));
+    privateStateKeyValueStorage
+        .updater()
+        .putPrivacyGroupHeadBlockMap(blockHeader.getHash(), privacyGroupHeadBlockHash)
```
There is no need to store an empty map!  You could store this map once all PMTs have been executed, but it should only happen if the map is not empty!

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/migration/PrivateStorageMetadataMigration.java:
```diff
@@ -0,0 +1,290 @@
+            .orElseThrow();
+
+    LOG.trace(
+        "Executed private transaction {} (result = {}}",
```
s/Executed/Simulated/

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/migration/PrivateStorageMetadataMigration.java:
```diff
@@ -0,0 +1,290 @@
+    privateBlockMetadata.addPrivateTransactionMetadata(privateTxMetadata);
+
+    updater.putPrivateBlockMetadata(
+        Bytes32.wrap(blockHeader.getHash()), privacyGroupIdBytes, privateBlockMetadata);
```
Do you really have to wrap that? I thought that a hash is a Bytes32 already ...

AbdelStark committed:
Handling of sequential memory modifications. (#316)

* Handling of sequential memory modifications.

* Handling of integer overflow.

* Handling of `MSTORE` opcode.
Added special case if exceptional halt reasons set is not empty.
Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

shemnon committed:
Add Block Body to Block Propagation Listener Context (#320)

Create plugin APIs for the block body (just transactions and ommers) and
add the block body to the block propagation listener context.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

macfarla committed:
[MINOR] americanize spelling (#322)

* americanize spelling

Signed-off-by: Sally MacFarlane <sally.macfarlane@consensys.net>

shemnon committed:
Generate trace test blocks.bin from json source (#323)

Generate the blocks.bin file from the json source.  The approach used is
to just run the besu import commands as a part of the build script.
This removes a magical binary file from the test files that is now
generated from human readable json.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

shemnon committed:
Add Parity style Statediff (#326)

* State diff

All but balances work at this point.  And it has something to do with
upfront gas costs.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* fix the balances bug

There was also a chained transactions bug. To address this I create a
new WorldUpdater for each iteration of a transaction through the block.
 The current world updater has the new state, the prior the old one.
 Diffs just fall out after that.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* get rid of midBlock state, it's unneeded with the chained updaters.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* remove rootWorld method, old code from a prior attempt.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* remove println

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* spotless

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* spotless, comments, and cleanup

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* spotless keeps missing this.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* javadoc

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* fix broken reference tests

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* spotless

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* tighten access

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

shemnon committed:
Fix zip building (#327)

* Fix zip build

Currently building from a non-git checkout, such as a zip file, does not
work because of the way we calculate the git commit hash.  Add a catch
clause to return 8 "x"s in case the calculation fails.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* let the user know retrievial failed

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

* left in debugging code tripping the error

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

AbdelStark committed:
[PIE-2224] Tracing API - Handling of contracts that produce logs. (#328)

* Handling of contracts that produce logs.

Added smart contracts to produce log records.

Fixes https://pegasys1.atlassian.net/browse/PIE-2224

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

shemnon committed:
Tracing cleanup and fixes (#331)

- Flat trace
  * Support all call types
  * Correct input for new tests
  * Correctly report out of gas
  * Report null result when no result but no error
- Remove need to keep messageFrame directly in traceframe
- Clean up StateDiffGenerator style
- Tighten type of updated accounts

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

AbdelStark committed:
[PIE-2268] Tracing API - Add tests specs to cover all options at the same time. (#332)



* Add tests with `trace`, `vmTrace` and `stateDiff` options at the same time.

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

rolfyone committed:
[BESU-182] Add tests to FilterParameter to demonstrate loading topics from JSON (#333)

* Add tests to FilterParameter to demonstrate loading topics from JSON

Demonstrate that filter topics in JSON format consistent with eth_newFilter RPC requests
will be correctly loaded into the FilterParameter topics.

Previous tests were loading from an internal structure, which makes it difficult to see
coverage of scenarios.

 - [] - matches anything

 - [A] matches A in the first position, and anything after

 - [A, B] - A in first position, and B in second position, then anything after

 - [null, B] - matches anything in first position, and B in second position, and anything after.

 - [[A,B],[A,B]] - Matches A OR B in first position, and A OR B in second position, and anything after.

Fixes Jira: [BESU-182]
Signed-off-by: Paul Harris <paul.harris@consensys.net>

Squash and merge
Signed-off-by: Sally MacFarlane <sally.macfarlane@consensys.net>

pinges committed:
Make priv_call use the PrivacyController (#325)

* Make priv_call use the PrivacyController

Signed-off-by: Stefan Pingel <stefan.pingel@consensys.net>

* minor changes after review

Signed-off-by: Stefan Pingel <stefan.pingel@consensys.net>

shemnon committed:
Trace APIs - fix reused blocktracer (#335)

Now that BlockTracer has mutable state we cannot re-use it across
requests, instead we need to create a new one for each call.

To support this spec tests were upgraded to allow batched requests and
responses, which are processed sequentially.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

AbdelStark committed:
[PIE-2223] Tracing API - Handling of various exceptional halt cases. (#334)


* Handling of various exceptional halt cases.

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

atoulme committed:
Fix debug logging shadowed by parameter (#319)

Signed-off-by: Antoine Toulme <antoine@lunar-ocean.com>

usmansaleem committed:
Use temporary self signed certificates for JSON-RPC HTTP TLS unit testing (#311)

 -- SelfSignedP12Certificate in testSupport for JSON-RPC HTTP TLS testing
 -- Generate self signed certificates in PKCS12 format with SAN extensions for testing
 -- Remove pre-generated key stores and trust stores used by TLS tests

Signed-off-by: Usman Saleem <usman@usmans.info>

shemnon committed:
Tracing - respect tx boundary for EIP-2200 cost calculations (#339)

When we calculate the cost of a SSTORE in trace we need to consider the
original value as of when the last transaction finished, not when the
whole block started.

To create a test for this the test blockchain also needed to be upgraded
to Istanbul.  This resulted in a large number of cost only changes.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

shemnon committed:
flat trace - swap and dup ordering (#336)

Swap and dup were getting out of order on the flat trace

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

usmansaleem committed:
[BESU-185] - CLI option to enable TLS client auth for JSON-RPC HTTP (#340)

Following cli options are added to enable TLS client authentication and trusting client certificates.

--rpc-http-tls-client-auth-enabled - Enable TLS client authentication for the JSON-RPC HTTP service (default: false)
--rpc-http-tls-known-clients-file - Path to file containing client's certificate common name and fingerprint for client authentication.
--rpc-http-tls-ca-clients-enabled - Enable to accept clients certificate signed by a valid CA for client authentication (default: false)

If client-auth is enabled, then user must either enable CA signed clients OR provide a known-clients file. An error is reported if both CA signed clients is disabled and known-clients file is not specified.

Signed-off-by: Usman Saleem <usman@usmans.info>

AbdelStark committed:
Add thank you to YourKit to the main project page. (#341)

* In order to obtain free open source YourKit license we are asked to put the link to YourKit website.

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

matkt committed:
[BESU-179] getSignerMetrics doesn't return missing signers (#343)

* fix signer metrics add missing validator

* fix end block parameter issue

Signed-off-by: Karim TAAM <karim.t2am@gmail.com>

shemnon committed:
Reduce Trace Memory Consumption (#344)

Switch the EVM Memory to a copy on write structure, so when the trace
reads the memory they don't have to copy mutable bytes but can instead
simply use the value as immutable.

Also, fix a state diff bug when an account is deleted that was also just
created in the TX don't report it as an all equal diff, as well as new accounts storage diffs.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

faraggi committed:
Refactor/contributor content  (#321)

* Contributor Content: complete refactor

- Deleted contributor files now hosted in wiki,
(CLI-STYLE-GUIDE.md, CODE-OF-CONDUCT.md, CODING-CONVENTIONS.md,
ROADMAP.md, SECURITY.md).
- Deleted old docs files in /docs folder (except readme for redirect).
- Refactored the root README.md with relevant content.
- Refactored the CONTRIBUTING.md file to simplify it with links to wiki.

* added contact developer advocate link
* small updates on README and CONTRIBUTING MDs.
* README updated.
* unified links for docs, jira, etc
* updated suggestions from PR comments
* updated links from comments
* updated links that still pointed to internal /docs
* LF wording and links

Signed-off-by: Felipe Faraggi <felipefaraggi@gmail.com>
Co-authored-by: Nicolas MASSART <NicolasMassart@users.noreply.github.com>

usmansaleem committed:
changelog 1.4 beta 3 (#345)

Signed-off-by: Usman Saleem <usman@usmans.info>

Co-authored-by: Tim Beiko <t.beiko23@gmail.com>

shemnon committed:
Mark Plugins Data and Events as stable (#346)

* Mark Plugins Data and Events as stable

Remove the @Unstable annotations form the data and BesuEvents services.
This means that no breaking changes (like removing methods or
changing signatures) will occur without proper notice and versioning
bumps.

* tweak build check

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

Co-authored-by: Tim Beiko <t.beiko23@gmail.com>

shemnon committed:
1.4.0-beta3 release (#347)

Signed-off-by: Edward Evans <edward.joshua.evans@gmail.com>

EdJoJob committed:
Prepare for version 1.4.0-beta4-SNAPSHOT (#348)

Signed-off-by: Edward Evans <edward.joshua.evans@gmail.com>

lucassaldanha committed:
Private state update metadata and migration

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>

shemnon committed:
Tracing multiple levels of calls now works properly (#352)

The iterative look back mechanism flat trace was using was missing
many of the nesting subtlties, so a stack based calculation is used.

A new test case with two levels of calls is also introduced.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

MadelineMurray committed:
Adding Danno's changes back (#355)

Signed-off-by: Madeline <madeline.murray@consensys.net>

macfarla committed:
[MINOR] typo (#357)

* typo

Signed-off-by: Sally MacFarlane <sally.macfarlane@consensys.net>

rolfyone committed:
[BESU-184] Update PicoCLI to latest version (#349)

- Updated PicoCLI from version 3.9.6 to 4.1.4 and fixed test cases that were impacted by breaking changes of PicoCLI.

- Updated the command factory to fallback to the default factory, as directed in the PicoCli v.4.0.0 release notes.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

shemnon committed:
Handle CREATE and CREATE2 in flat trace (#353)

* Handle CREATE and CREATE2 in flat trace

Handle create and create2 operations. Some other tests also got adjusted
as the source accounts were running out of test ether.  Bumping up the
value in the genesis caused the statediff values to also go up.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

AbdelStark committed:
[BESU-176] JRPC response to eth_getTransactionCount is different from other node implementations. (#351)

* Fix strategy for getting next nonce for sender in `PendingTransactions`.
Added test.

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

* Introduced `TransactionsForSenderInfo` structure to have a less naive implementation, more robust against load testing.
A queue of gaps between used nonces is maintained.
`getNextNonceForSender` has been updated and have the following logic:
- if no infos corresponding to the nonce returns an empty optional
- if gap list is not empty returns the lowest nonce in the gap
- else returns the max nonce + 1

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

* spotless

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

* spotless

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

* added test

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

* spotless

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

lucassaldanha committed:
Fix migration of failed txs

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>

joshuafernandes committed:
add the binary on ci at build time, dont need it to be in the code base (#359)

Signed-off-by: Joshua Fernandes <joshua.fernandes@consensys.net>

macfarla committed:
[BESU-80] TLS to orion (#324)

* enable SSL on connection to enclave

Signed-off-by: Sally MacFarlane <sally.macfarlane@consensys.net>

lucassaldanha committed:
Remove restriction for priv_call

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>

lucassaldanha committed:
Merge branch 'master' into updated-migration

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>

AbdelStark committed:
Renamed .index to .cache. (#364)

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

shemnon committed:
Json Blocks Import timestamp (#358)

Allows for the setting of the timestamp in JSON block importer.
Ordinarily it is now and then 1 second for each block.  The initial time
can now be set via CLI option `--start-time`, which defaults to now.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

lucassaldanha committed:
Fix import

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>

shemnon committed:
More tracing api fixes (#361)

* Handle CREATE and CREATE2 in flat trace
* precompile calls don't generate traces
* Handle REVERT in output and result nodes.
* output doesn't need to be held behind a supplier
* Handle padded output values when end is past memory
* Store memory in vm trace for MLOADs
* CREATE and CREATE2 store memory in vm-trace

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

lucassaldanha committed:
Merge branch 'master' into updated-migration

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>

lucassaldanha committed:
Fix unit test

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>

lucassaldanha committed:
Changed migration msg

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>

edwardmack committed:
[ETC] Aztlan fork support (#366)

This PR introduces functionality to conform with ECIP-1061
https://ecips.ethereumclassic.org/ECIPs/ecip-1061

Signed-off-by: edwardmack <ed@edwardmack.com>
Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

Co-authored-by: Danno Ferrin <danno.ferrin@shemnon.com>

pinges committed:
Merge branch 'master' into updated-migration

shemnon committed:
Plugin error stack traces (#369)

Because of how the Log4J2 api works exception stack traces were not
being printed.  Update to use the explicit "throwable" overloaded
methods.

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

shemnon committed:
VM Trace fixes (#372)

* correct refund addresses
* correct returned memory from static precompiled calls.
* update integration test
* precompiles sometimes get plain old CALLs

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

lucassaldanha committed:
Merge branch 'master' into updated-migration

Signed-off-by: Lucas Saldanha <lucas.saldanha@consensys.net>
